tools/integration-test/src/tests/ica_filter.rs

```diff
@@ -0,0 +1,266 @@
+    run_binary_chain_test(&IcaFilterTestDeny)
+}
+
+pub struct IcaFilterTestDeny;
+
+impl TestOverrides for IcaFilterTestDeny {
+    // Use deterministic identifiers for clients, connections, and channels
+    fn modify_test_config(&self, config: &mut TestConfig) {
+        config.chain_command_path = "icad".to_string();
+        config.bootstrap_with_random_ids = false;
+    }
+
+    // Enable channel workers and deny ICA ports
+    fn modify_relayer_config(&self, config: &mut Config) {
+        config.mode.channels.enabled = true;
+
+        for chain in &mut config.chains {
+            chain.packet_filter = PacketFilter::Deny(ChannelFilters::new(vec![(
+                FilterPattern::Wildcard("ica*".parse().unwrap()),
+                FilterPattern::Wildcard("*".parse().unwrap()),
+            )]));
+        }
+    }
+}
+
+impl BinaryChainTest for IcaFilterTestDeny {
+    fn run<ChainA: ChainHandle, ChainB: ChainHandle>(
+        &self,
+        _config: &TestConfig,
+        relayer: RelayerDriver,
+        chains: ConnectedChains<ChainA, ChainB>,
+    ) -> Result<(), Error> {
+        // Register an interchain account on behalf of controller wallet `user1`
+        // where the counterparty chain is the interchain accounts host.
+        // Then spawn the supervisor.
+        let (_handle, _, _, channel_id_a, port_id_a) =
+            bootstrap_and_register_interchain_account(&relayer, &chains)?;
+
+        // Wait a bit, the relayer will refuse to complete the channel handshake
+        // because the port is explicitly disallowed by the filter.
+        std::thread::sleep(Duration::from_secs(30));
+
+        let channel_end = query_channel_end(
+            chains.handle_a(),
+            &channel_id_a.as_ref(),
+            &port_id_a.as_ref(),
+        )?;
+
+        // Check that the channel is left in state Init
+        assert_eq(
+            "channel end should still be in state Init",
+            channel_end.value().state(),
+            &State::Init,
+        )
+    }
+}
+
+#[allow(clippy::type_complexity)]
+fn bootstrap_and_register_interchain_account<ChainA: ChainHandle, ChainB: ChainHandle>(
+    relayer: &RelayerDriver,
+    chains: &ConnectedChains<ChainA, ChainB>,
+) -> Result<
+    (
+        SupervisorHandle,
+        MonoTagged<ChainA, Wallet>,
+        ConnectedConnection<ChainA, ChainB>,
+        TaggedChannelId<ChainA, ChainB>,
+        TaggedPortId<ChainA, ChainB>,
+    ),
+    Error,
+> {
+    let connection_a =
+        bootstrap_connection(&chains.foreign_clients, default_connection_delay(), false)?;
+
+    let wallet_a = chains.node_a.wallets().user1().cloned();
+
+    let handle = relayer.spawn_supervisor()?;
+
+    chains
+        .node_a
+        .chain_driver()
+        .register_interchain_account(&wallet_a.address(), &connection_a.connection_id_a.as_ref())?;
+
+    let channel_id_a: TaggedChannelId<ChainA, ChainB> =
+        TaggedChannelId::new("channel-0".parse().unwrap());
```

Comment:
Is there any way we can get back the channel ID that is not hard coded? Should Hermes also implement CLI and function for registering interchain account instead of doing it through Gaia?

Reply:
> Should Hermes also implement CLI and function for registering interchain account instead of doing it through Gaia?

I don't think that should be the responsibility of the relayer, though it might be useful for testing but it's not clear whether it's worth the cost or not at the moment. Maybe if we hear the need from operators.

We can discuss during the weekly sync, but let's keep it out of this PR anyway.